doc/asic-fw-update/asic-fw-update-config.md

```diff
@@ -0,0 +1,343 @@
+
+### High-Level Design
+#### Original ASIC FW update flow (Nvidia)
+The original flow of ASIC firmware update is very simple. Every time system
+reboot causes execution of **platform_reboot** script. This is platform specific
+script that prepares system for the reboot. One of the responsibility of that
+script is to make sure that ASIC firmware is up-to-date. To perform that check
+it calls **mlnx-fw-upgrade** script (for Nvidia platform) which actually
+performs checking of the versions (available and installed). In case versions
+are different it installs the ASIC firmware embed into system image
+(form */etc/mlnx/* folder for Nvidia platform). See [Figure 1. ASIC firmware
+update original flow](#figure-1-asic-firmware-update-original-flow).
+
+<!-- omit in toc -->
+###### Figure 1. ASIC firmware update original flow (Nvidia)
+<p align=center>
+  <img src="img/asic-firmware-update-orig-sequence.png"
+       alt="Figure 1. ASIC firmware update original flow">
+</p>
+
+When updating through ONIE **start.sh** script of **syncd** container performs
+ASIC firmware installation. It calls the same **mlnx-fw-upgrade** script in
+order to make sure that ASIC firmware is up-to-date.
+
+#### New ASIC FW update flow
+The new ASIC firmware update flow brings a couple of improvements to ASIC
+firmware update script. It will read the configuration from the database and
+based on it, decide if ASIC FW have to be updated.
+In the same manner, the script will know the source of the firmware image:
+system image or user defined.
+Example: for Nvidia platform system image, ASIC firmware image located in
+*/etc/mlnx/* directory. We bring a new user location for ASIC firmware image
+*/host/mlnx/asic/*. This location will survive system upgrade, so system will be
+able to update ASIC FW even after system upgrade. See the new flow here:
+[Figure 2. ASIC firmware update controlled flow
+](#figure-2-asic-firmware-update-controlled-flow).
+
+<!-- omit in toc -->
+###### Figure 2. ASIC firmware update controlled flow
+<p align=center>
+  <img src="img/asic-firmware-update-new-sequence.png"
+       alt="Figure 2. ASIC firmware update controlled flow">
+</p>
+
+### Configuration and management
+#### CLI Enhancements
+The initial version of that feature brings an ability to configure via
+config_db.json, but not a CLI commands.
+But based on the current SONiC CLI tree, we can propose the next commands tree.
+
+<!-- omit in toc -->
+#### CLI
+Proposed ASIC firmware control CLI tree:
+```
+config
+\-- firmware
+    \-- asic
+        |-- auto-apdate [ enable | disable ]
+        \-- default [ image | user ]
+```
+
+CLI helptext:
+```
+admin@sonic:~$ sudo config firmware
+Usage: config firmware asic [OPTIONS] COMMAND [ARGS]...
+
+  AAA command line
+
+Options:
+  -h, -?, --help  Show this message and exit.
+
+Commands:
+  auto-update     Control ASIC firmware auto update [ enable | disable ]
+  default         Set default firmware source [ image | user ]
```

Comment:
> Instead of default can't we name it as 'source'?

We used 'default' in our Nvidia code, so I prefer to leave it as it is.

> Does 'user' option is fixed at /host/mlnx/asic/. In that case can it be part of CLI help and info message when user configures this option?

Yes, it is. We can include it in the CLI helptext

> Additionally when user option is specified for source and the user path doesn't exist should there be error?

For Nvidia platforms, `mlnx-fw-upgrade` script will notify the user about non-existent files.